libgnucash/engine/Transaction.c

```diff
@@ -2472,15 +2481,18 @@ xaccTransGetTxnType (const Transaction *trans)
 {
     const char *s = NULL;
     GValue v = G_VALUE_INIT;
+    char ret = TXN_TYPE_NONE;
 
     if (!trans) return TXN_TYPE_NONE;
     qof_instance_get_kvp (QOF_INSTANCE (trans), &v, 1, TRANS_TXN_TYPE_KVP);
     if (G_VALUE_HOLDS_STRING (&v))
+    {
          s = g_value_get_string (&v);
-    if (s && strlen (s) == 1)
-        return *s;
-
-    return TXN_TYPE_NONE;
+         if (s && (s[0] != '\0') && (s[1] == '\0'))
```

Comment:
Why this change ?

Reply:
Small micro-optimisation. `strlen` is a function call and is O(N), whereas this translates to 6 assembly instructions...

Reply:
Fair enough.

On the other hand the original code expresses the intent much more clearly where your version requires mental decoding to figure out we are trying the first character of a one-byte string, only if the string is exactly one byte. So unless you have measurable performance gains from this I would favour the more expressive form.

Just a nitpick though.

Reply:
Indeed.

We're not in C++ land yet, but would there be a C++ alternative not using macros ? I hear a lot of backpush against macros in C++ conversations.

Reply:
Sort of, and it works in C too:
```
static inline char 
str_one_char (const char* s) 
{
   return (s && s[0] && !s[1]) ? *s : '\0'; 
}
```
like a macro the *definition* has to be part of the compilation unit (so it's either in the source file being compiled or in a header) or it can't be inlined. Also `inline` is a suggestion, the compiler doesn't have to inline it and probably won't if told `-O0` and if it doesn't inline it then the string has to be pretty long before it improves on `strlen`.

Reply:
let's resolve this. `strlen` it is then.